base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
+               ::Val{CHUNK_SIZE}, t::AbstractVector{U}) where {U <: Unsigned, CHUNK_SIZE}
+    # bits is unsigned and CHUNK_SIZE is a compile time constant for performance reasons.
+    MASK = UInt(1) << CHUNK_SIZE - 0x1
+    counts = Vector{UInt}(undef, MASK+2)
+
+    @inbounds for shift in 0:CHUNK_SIZE:bits-1
+
+        # counts[2:MASK+2] will store the number of elements that fall into each bucket.
+        # if CHUNK_SIZE = 8, counts[2] is bucket 0x00 and counts[257] is bucket 0xff.
+        counts .= 0
+        for k in lo:hi
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 2 # compute its bucket's index for this pass
+            counts[i] += 1            # increment that bucket's count
+        end
+
+        counts[1] = lo                # set target index for the first bucket
+        cumsum!(counts, counts)       # set target indices for subsequent buckets
+        # counts[1:MASK+1] now stores indices where the first member of each bucket
+        # belongs, not the number of elements in each bucket. We will put the first element
+        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
+        # and the last element of bucket 0x00 in t[counts[2]-1].
+
+        for k in lo:hi # Is this iteration slower than it could be?
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 1 # compute its bucket's index for this pass
+            j = counts[i]             # lookup the target index
+            t[j] = x                  # put the element where it belongs
+            counts[i] = j + 1         # increment the target index for the next
+        end                           #  ↳ element in this bucket
+
+        v, t = t, v # swap the now sorted destination list t back into the primary list v
+
+    end
+
+    v
+end
+
+function sort!(v::AbstractVector{<:Bool}, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    first = lt(o, false, true) ? false : lt(o, true, false) ? true : return v
+    count = 0
+    for i in lo:hi
+        if v[i] == first
+            count += 1
+        end
+    end
+    v[lo:lo+count-1] .= first
+    v[lo+count:hi] .= !first
+    v
+end
+
+maybe_unsigned(x::Integer) = x # this is necessary to avoid calling unsigned on BigInt
+maybe_unsigned(x::Union{Int8, Int16, Int32, Int64, Int128}) = unsigned(x)
+function _extrema(v::AbstractArray, lo::Integer, hi::Integer, o::Ordering)
+    mn = mx = v[lo]
+    while lo < hi
+        lo += 1
+        vi = v[lo]
+        lt(o, vi, mn) && (mn = vi)
+        lt(o, mx, vi) && (mx = vi)
+    end
+    mn, mx
+end
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::AdaptiveSort, o::Ordering)
+    # if the sorting task is unserializable, then we can't radix sort or sort_int_range!
+    # so we skip straight to the fallback algorithm which is comparison based.
+    U = Serial.Serializable(eltype(v), o)
```

Comment:
this seems very weird.

Reply:
`Sort.Serial.Serializable` is a trait that indicates both whether something is serializable and to what datatype it serializes to. For example, by default, sorting by the function `f(x) = x & 0xff` is serializable, but if someone specialized `Base.Sort.Serial.serialize` and `Base.Sort.Serial.deserialize` for `f`, then they could also specialize `Base.Sort.Serial.Serializable` to return `UInt8`.

AdaptiveSort uses this trait to determine if any of its optimizations are possible.

Reply:
I also find that "serialize" could be confusing as it's completely different from `Base.serialize`. Maybe we could find another term, which more explicitly indicates that this maps a value to an unsigned integer?

Reply:
SortingAlgorithms.jl uses `uint_mapping`. A key facet of this implementation is the mapping is reversible, which led me to the term serialization. I'm not a fan of `UIntBijection`.

Reply:
Perhaps `Serial` -> `UIntMapping`, `serialize` -> `uint_map`, `deserialize` -> `uint_unmap`, `Serializable` -> `UIntMappable`?

Reply:
that sounds good

Reply:
+1

Maybe these don't need to go into a submodule BTW.

Reply:
The functions do all have "uint_map" in their name, so there shouldn't be an issue with taking them out of a submodule. I was following the example of `Float`, but `Float` defines things like `Left` and `specials_to_end` which are once to keep in a Submodule and `Serial`/`UIntMapping` does not define this sort of thing.

Reply:
Renamed and removed the submodule.